src/haz3lcore/dynamics/TypeAssignment.re

```diff
@@ -0,0 +1,288 @@
+         })
+      |> OptUtil.sequence;
+
+    switch (typ_cases) {
+    | None => None
+    | Some(_) => Some(Typ.Unknown(Internal))
+    };
+  | Closure(_, d) => typ_of_dhexp(ctx, m, d)
+  | Filter(_, d) => typ_of_dhexp(ctx, m, d)
+  | BoundVar(name) =>
+    let+ var = Ctx.lookup_var(ctx, name);
+    var.typ;
+  | Sequence(d1, d2) =>
+    let* _ = typ_of_dhexp(ctx, m, d1);
+    typ_of_dhexp(ctx, m, d2);
+  | Let(dhp, de, db) =>
+    let* ty1 = typ_of_dhexp(ctx, m, de);
+    typ_of_dhexp(dhpat_extend_ctx(dhp, ty1, ctx), m, db);
+  | FixF(name, ty1, d) =>
+    let entry = Ctx.VarEntry({name, id: Id.invalid, typ: ty1});
+    typ_of_dhexp(Ctx.extend(ctx, entry), m, d);
+  | Fun(dhp, ty1, d, _) =>
+    let+ ty2 = typ_of_dhexp(dhpat_extend_ctx(dhp, ty1, ctx), m, d);
+    Typ.Arrow(ty1, ty2);
+  | Ap(d1, d2) =>
+    let* ty1 = typ_of_dhexp(ctx, m, d1);
+    let* ty2 = typ_of_dhexp(ctx, m, d2);
+    switch (arrow_aux(ty1)) {
+    | Arrow(tyl, tyr) when Typ.eq(tyl, ty2) => Some(tyr)
+    | _ => None
+    };
+  | ApBuiltin(_)
+  | BuiltinFun(_) => None
+  | Test(_, dtest) =>
+    let* ty = typ_of_dhexp(ctx, m, dtest);
+    Typ.eq(ty, Bool) ? Some(Typ.Prod([])) : None;
+  | BoolLit(_) => Some(Bool)
+  | IntLit(_) => Some(Int)
+  | FloatLit(_) => Some(Float)
+  | StringLit(_) => Some(String)
+  | BinBoolOp(_, d1, d2) =>
+    let* ty1 = typ_of_dhexp(ctx, m, d1);
+    let* ty2 = typ_of_dhexp(ctx, m, d2);
+    Typ.eq(ty1, Bool) && Typ.eq(ty2, Bool) ? Some(Typ.Bool) : None;
+  | BinIntOp(op, d1, d2) =>
+    let* ty1 = typ_of_dhexp(ctx, m, d1);
+    let* ty2 = typ_of_dhexp(ctx, m, d2);
+    if (Typ.eq(ty1, Int) && Typ.eq(ty2, Int)) {
+      switch (op) {
+      | Minus
+      | Plus
+      | Times
+      | Power
+      | Divide => Some(Typ.Int)
+      | LessThan
+      | LessThanOrEqual
+      | GreaterThan
+      | GreaterThanOrEqual
+      | Equals
+      | NotEquals => Some(Typ.Bool)
+      };
+    } else {
+      None;
+    };
+  | BinFloatOp(op, d1, d2) =>
+    let* ty1 = typ_of_dhexp(ctx, m, d1);
+    let* ty2 = typ_of_dhexp(ctx, m, d2);
+    if (Typ.eq(ty1, Float) && Typ.eq(ty2, Float)) {
+      switch (op) {
+      | Minus
+      | Plus
+      | Times
+      | Power
+      | Divide => Some(Typ.Float)
+      | LessThan
+      | LessThanOrEqual
+      | GreaterThan
+      | GreaterThanOrEqual
+      | Equals
+      | NotEquals => Some(Typ.Bool)
+      };
+    } else {
+      None;
+    };
+  | BinStringOp(op, d1, d2) =>
+    let* ty1 = typ_of_dhexp(ctx, m, d1);
+    let* ty2 = typ_of_dhexp(ctx, m, d2);
+    if (Typ.eq(ty1, String) && Typ.eq(ty2, String)) {
+      switch (op) {
+      | Concat => Some(Typ.String)
+      | Equals => Some(Typ.Bool)
+      };
+    } else {
+      None;
+    };
+  | ListLit(_, _, ty, _) => Some(List(ty))
+  | Cons(d1, d2) =>
+    let* ty1 = typ_of_dhexp(ctx, m, d1);
+    let* ty2 = typ_of_dhexp(ctx, m, d2);
+    switch (ty2) {
+    | List(Unknown(Internal)) => Some(Typ.List(ty1))
```

Comment:
we don't need to do any consistency things here -- the head element should have the same type as the element type of the tail, which must be a list type.